File: go/kdb/integrity/verify_test.go
package integrity

import (
	"testing"

	"github.com/limidus/kdb/go/kdb/storage"
)

func TestVerifyCleanLogPassesAllLevels(t *testing.T) {
	shim := newTestShim(t)
	const ns = "ns1"
	commits := buildChain(t, 3, ns)
	appendSegment(t, shim, ns, 0, rawFrame(t, commits[0]), rawFrame(t, commits[1]), rawFrame(t, commits[2]))

	report, err := Verify(shim, ns, Options{Level: L2, Compression: storage.CompressionNone})
	if err != nil {
		t.Fatal(err)
	}
	if !report.Clean() {
		t.Fatalf("expected clean report, got findings: %+v", report.Findings)
	}
	if len(report.Segments) != 1 || report.Segments[0].FrameCount != 3 {
		t.Fatalf("expected 1 segment with 3 frames, got %+v", report.Segments)
	}
}

func TestVerifyDetectsMidLogCorruption(t *testing.T) {
	shim := newTestShim(t)
	const ns = "ns1"
	commits := buildChain(t, 3, ns) // c0 <- c1 <- c2
	appendSegment(t, shim, ns, 0, rawFrame(t, commits[0]), flippedFrame(t, commits[1]))
	appendSegment(t, shim, ns, 1, rawFrame(t, commits[2]))

	report, err := Verify(shim, ns, Options{Level: L2, Compression: storage.CompressionNone})
	if err != nil {
		t.Fatal(err)
	}
	if report.Clean() {
		t.Fatal("expected findings, got a clean report")
	}
	if !hasFinding(report, MidLogCorruption, 0) {
		t.Errorf("expected mid_log_corruption at segment 0, got %+v", report.Findings)
	}
	if !hasFindingWithHash(report, MissingParent, commits[1].Hash.Hex()) {
		t.Errorf("expected missing_parent for the corrupted commit %s, got %+v", commits[1].Hash.Hex(), report.Findings)
	}
}

func TestVerifyCrcMismatchOnLastSegmentIsTornTail(t *testing.T) {
	shim := newTestShim(t)
	const ns = "ns1"
	commits := buildChain(t, 2, ns)
	// Single segment, so any corruption in it is corruption in the last
	// segment - kdb-spec-layer13 §4.3 classifies this as a torn tail
	// regardless of *why* the CRC failed, not just outright truncation.
	appendSegment(t, shim, ns, 0, rawFrame(t, commits[0]), flippedFrame(t, commits[1]))

	report, err := Verify(shim, ns, Options{Level: L1, Compression: storage.CompressionNone})
	if err != nil {
		t.Fatal(err)
	}
	if len(report.Findings) != 1 || report.Findings[0].Classification != TornTail {
		t.Fatalf("expected exactly one torn_tail finding, got %+v", report.Findings)
	}
}

func TestVerifyDetectsTornTail(t *testing.T) {
	shim := newTestShim(t)
	const ns = "ns1"
	commits := buildChain(t, 2, ns)
	full := rawFrame(t, commits[0])
	torn := rawFrame(t, commits[1])
	torn = torn[:len(torn)-5] // cut short: declared length no longer fits
	appendSegment(t, shim, ns, 0, full, torn)

	report, err := Verify(shim, ns, Options{Level: L2, Compression: storage.CompressionNone})
	if err != nil {
		t.Fatal(err)
	}
	if len(report.Findings) != 1 {
		t.Fatalf("expected exactly one finding, got %+v", report.Findings)
	}
	f := report.Findings[0]
	if f.Classification != TornTail || f.Offset != len(full) {
		t.Fatalf("expected torn_tail at offset %d, got %+v", len(full), f)
	}
	if report.Segments[0].FrameCount != 1 {
		t.Fatalf("expected only the clean frame counted, got %+v", report.Segments[0])
	}
}

func TestVerifyDetectsSequenceGap(t *testing.T) {
	shim := newTestShim(t)
	const ns = "ns1"
	commits := buildChain(t, 2, ns)
	appendSegment(t, shim, ns, 0, rawFrame(t, commits[0]))
	appendSegment(t, shim, ns, 2, rawFrame(t, commits[1])) // sequence 1 is missing

	report, err := Verify(shim, ns, Options{Level: L1, Compression: storage.CompressionNone})
	if err != nil {
		t.Fatal(err)
	}
	if !hasFinding(report, SequenceGap, 2) {
		t.Fatalf("expected sequence_gap at segment 2, got %+v", report.Findings)
	}
}

func TestVerifyNeverMutatesOnDisk(t *testing.T) {
	shim := newTestShim(t)
	const ns = "ns1"
	commits := buildChain(t, 2, ns)
	appendSegment(t, shim, ns, 0, rawFrame(t, commits[0]), rawFrame(t, commits[1]))

	before, err := shim.ReadFromSegment("ns/"+ns+"/delta/00000000000000000000.seg", 0, 1<<20)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := Verify(shim, ns, Options{Level: L2, Compression: storage.CompressionNone}); err != nil {
		t.Fatal(err)
	}
	after, err := shim.ReadFromSegment("ns/"+ns+"/delta/00000000000000000000.seg", 0, 1<<20)
	if err != nil {
		t.Fatal(err)
	}
	if string(before) != string(after) {
		t.Fatal("Verify must not mutate segment bytes")
	}
}

func hasFinding(report *Report, class Classification, segment int64) bool {
	for _, f := range report.Findings {
		if f.Classification == class && f.Segment == segment {
			return true
		}
	}
	return false
}

func hasFindingWithHash(report *Report, class Classification, hash string) bool {
	for _, f := range report.Findings {
		if f.Classification == class && f.CommitHash == hash {
			return true
		}
	}
	return false
}

File: kdb-integrity/src/commonTest/kotlin/dev/kdb/integrity/VerifyTest.kt
package dev.kdb.integrity

import dev.kdb.storage.CompressionCodec
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import kotlinx.coroutines.test.runTest

class VerifyTest {
    @Test
    fun cleanLogPassesAllLevels() =
        runTest {
            val shim = newTestShim()
            val ns = "ns1"
            val commits = buildChain(3, ns)
            appendSegment(shim, ns, 0, rawFrame(commits[0]), rawFrame(commits[1]), rawFrame(commits[2]))

            val report = verify(shim, ns, Options(Level.L2, CompressionCodec.NONE))
            assertTrue(report.isClean, "expected clean report, got ${report.findings}")
            assertEquals(1, report.segments.size)
            assertEquals(3, report.segments[0].frameCount)
        }

    @Test
    fun detectsMidLogCorruption() =
        runTest {
            val shim = newTestShim()
            val ns = "ns1"
            val commits = buildChain(3, ns) // c0 <- c1 <- c2
            appendSegment(shim, ns, 0, rawFrame(commits[0]), flippedFrame(commits[1]))
            appendSegment(shim, ns, 1, rawFrame(commits[2]))

            val report = verify(shim, ns, Options(Level.L2, CompressionCodec.NONE))
            assertFalse(report.isClean)
            assertTrue(hasFinding(report, Classification.MID_LOG_CORRUPTION, 0), "expected mid_log_corruption at segment 0, got ${report.findings}")
            assertTrue(
                hasFindingWithHash(report, Classification.MISSING_PARENT, commits[1].hash.toHex()),
                "expected missing_parent for ${commits[1].hash.toHex()}, got ${report.findings}",
            )
        }

    @Test
    fun crcMismatchOnLastSegmentIsTornTail() =
        runTest {
            val shim = newTestShim()
            val ns = "ns1"
            val commits = buildChain(2, ns)
            // Single segment, so any corruption in it is corruption in the last segment -
            // kdb-spec-layer13 §4.3 classifies this as a torn tail regardless of *why* the CRC
            // failed, not just outright truncation.
            appendSegment(shim, ns, 0, rawFrame(commits[0]), flippedFrame(commits[1]))

            val report = verify(shim, ns, Options(Level.L1, CompressionCodec.NONE))
            assertEquals(1, report.findings.size)
            assertEquals(Classification.TORN_TAIL, report.findings[0].classification)
        }

    @Test
    fun detectsTornTail() =
        runTest {
            val shim = newTestShim()
            val ns = "ns1"
            val commits = buildChain(2, ns)
            val full = rawFrame(commits[0])
            val torn = rawFrame(commits[1]).copyOfRange(0, 5) // cut short: declared length no longer fits
            appendSegment(shim, ns, 0, full, torn)

            val report = verify(shim, ns, Options(Level.L2, CompressionCodec.NONE))
            assertEquals(1, report.findings.size)
            val f = report.findings[0]
            assertEquals(Classification.TORN_TAIL, f.classification)
            assertEquals(full.size, f.offset)
            assertEquals(1, report.segments[0].frameCount)
        }

    @Test
    fun detectsSequenceGap() =
        runTest {
            val shim = newTestShim()
            val ns = "ns1"
            val commits = buildChain(2, ns)
            appendSegment(shim, ns, 0, rawFrame(commits[0]))
            appendSegment(shim, ns, 2, rawFrame(commits[1])) // sequence 1 is missing

            val report = verify(shim, ns, Options(Level.L1, CompressionCodec.NONE))
            assertTrue(hasFinding(report, Classification.SEQUENCE_GAP, 2), "expected sequence_gap at segment 2, got ${report.findings}")
        }

    @Test
    fun neverMutatesOnDisk() =
        runTest {
            val shim = newTestShim()
            val ns = "ns1"
            val commits = buildChain(2, ns)
            appendSegment(shim, ns, 0, rawFrame(commits[0]), rawFrame(commits[1]))

            val name = "ns/$ns/delta/00000000000000000000.seg"
            val before = shim.readFromSegment(name, 0, 1 shl 20)
            verify(shim, ns, Options(Level.L2, CompressionCodec.NONE))
            val after = shim.readFromSegment(name, 0, 1 shl 20)
            assertTrue(before.contentEquals(after), "verify must not mutate segment bytes")
        }
}

internal fun hasFinding(report: Report, classification: Classification, segment: Long): Boolean =
    report.findings.any { it.classification == classification && it.segment == segment }

internal fun hasFindingWithHash(report: Report, classification: Classification, hash: String): Boolean =
    report.findings.any { it.classification == classification && it.commitHash == hash }
